base/operators.jl

```diff
@@ -1332,22 +1332,41 @@ used to implement specialized methods.
 """
 in(x) = Fix2(in, x)
 
-for ItrT = (Tuple,Any)
-    # define a generic method and a specialized version for `Tuple`,
-    # whose method bodies are identical, while giving better effects to the later
-    @eval function in(x, itr::$ItrT)
-        $(ItrT === Tuple ? :(@_terminates_locally_meta) : :nothing)
-        anymissing = false
-        for y in itr
-            v = (y == x)
-            if ismissing(v)
-                anymissing = true
-            elseif v
-                return true
-            end
+function in(x, itr::Any)
+    anymissing = false
+    for y in itr
+        v = (y == x)
+        if ismissing(v)
+            anymissing = true
+        elseif v
+            return true
         end
+    end
+    return anymissing ? missing : false
+end
+
+# Specialized variant of in for Tuple, which can generate typed comparisons for each element
+# of the tuple, skipping values that are statically known to be != at compile time.
+function in(x, itr::Tuple)
+    @_terminates_locally_meta
+    _in_tuple(x, itr, false)
+end
+# This recursive function will be unrolled at compiletime, and will not generate separate
+# llvm-compiled specializations for each step of the recursion.
+@inline function _in_tuple(x, @nospecialize(itr::Tuple), anymissing::Bool)
+    @_terminates_locally_meta
```

Comment:
Does this `@_terminates_locally_meta` still apply? Should it go in both functions?

Reply:
It should be replaced by `@_terminates_globally` on the `in` method.

Reply:
and we can remove the annotation here

Reply:
What if `==` for the tuple elements doesn't terminate?

Reply:
Thanks, @aviatesk. I've applied your suggestion in [8968e5e](https://github.com/JuliaLang/julia/pull/54026/commits/8968e5ed5d0d04d5dcfb2ae3670372ea813c7a2d), by removing the annotation in the callee, and switching to the global annotation in the top-level function.

I have to say, @martinholters's comment does make sense... That seems like it was an issue before my PR, and is still an issue after?

Reply:
@martinholters is right, and now I think we can actually remove the `@_terminates_globally` annotation now since the compiler will be able to prove it automatically given the recursive implementation.